thirdparty/faiss/faiss/IndexIVFFlat.cpp

```diff
@@ -287,6 +300,38 @@ void IndexIVFFlatCC::reconstruct_from_offset_without_codes(
         float* recons) const {
     FAISS_THROW_MSG("ivfflat_cc index not support reconstruct_from_offset_without_codes operation");
 }
+
+void IndexIVFFlatCC::train(idx_t n, const float* x) {
+    const float* data_pointer;
+    std::unique_ptr<float[]> normal_buffers_;
+    if (is_cosine_) {
+        normal_buffers_ = std::make_unique<float[]>( n * d * sizeof(float));
+        std::memcpy(normal_buffers_.get(), x , n * d * sizeof(float));
+        knowhere::NormalizeVecs(normal_buffers_.get(), n, d);
+        data_pointer = normal_buffers_.get();
+    } else {
+        data_pointer = x;
+    }
+    IndexIVF::train(n, data_pointer);
+}
+
+void IndexIVFFlatCC::add_with_ids(idx_t n, const float* x, const idx_t* xids) {
+    std::unique_ptr<idx_t[]> coarse_idx(new idx_t[n]);
+    const float* assign_x;
+    std::unique_ptr<float[]> assign_buffer;
+    std::vector<float> normals;
+    if (is_cosine_) {
+        assign_buffer = std::make_unique<float[]>(n * d * sizeof(float));
+        std::memcpy(assign_buffer.get(), x, n * d * sizeof(float));
+        normals = knowhere::NormalizeVecs(assign_buffer.get(), n, d);
```

Comment:
better to set "normals" as an input parameter for NormalizeVecs
knowhere::NormalizeVecs(asign_buffer.get(), n, d, normals)